### PR TITLE
Use lockGuardService to prevent back navigation

### LINKED ISSUE #1882: Back button can navigate to lock page without vault being locked
## Describe the Bug
When unlocking the extension you can press your mouse's back button and go back to a seemingly locked vault, but if you press your mouse's forward button you'll go back to the unlocked vault.

Not a big deal but this behaviour looks a bit strange

## Steps To Reproduce

1. Open the extension
2. Provide you master password to unlock your vault
3. Press your mouse's back button, you'll be redirected to the previous panel and vault will look locked
4. Press your mouse's forward button, you'll be back into your vault (without having entered the master password)

## Expected Result

The mouse back button should not redirect you to the panel asking you for your master password

## Actual Result

You get redirected to the panel asking you for your master password, but you can still go back to your vault without providing the password (even though a message on the panel says that your vault is locked)

## Screenshots or Videos

bitwarden browser

## Environment

- Operating system: Ubuntu 20.04.2 LTS
- Browser: Chromium Version 91.0.4472.77 (Official Build) snap (64-bit)
- Build Version: Version: 1.50.0

## Additional Context

This does not happen in Firefox, there the mouse back button doesn't redirect you to the master password panel

The mouse events are button 8 press and 9 press for back and forward respectively

If you close the extension and reopen it the back button does not redirect you to the master password panel (basically it happens only after you've just accessed your vault without closing the extension)

At some point this issue was also present in the desktop version but I believe it is not longer there (https://github.com/bitwarden/desktop/issues/314)

### PR DESCRIPTION
**Pull request by eliykat**
## Objective

Fix #1882:
> When unlocking the extension you can press your mouse's back button and go back to a seemingly locked vault, but if you press your mouse's forward button you'll go back to the unlocked vault.

I fixed this in web and desktop by adding `lockGuardService`, which redirects to the vault if the user is already unlocked (https://github.com/bitwarden/desktop/pull/847), so I used the same here.

Bonus refactor: this bug doesn't affect the login page, because that's already protected by `launchGuardService`. However, `launchGuardService` duplicates a lot of logic from jslib's `unauthService`, so I have removed the duplicate code while I'm here. This means that all Angular clients use the jslib route guards rather than having their own implementations.

## Code changes

* Add `lockGuardService` to the `lock` route. This is subclassed so that we can override the redirect route if the vault is authenticated & unlocked - desktop and web use the `vault` route as the landing page on successful login, whereas browser uses `tabs/current`. I'm not sure whether this is strictly necessary, but it felt like unexpected behaviour when I was testing it without making this change.

* Extract the duplicate logic from `launchGuardService` and call `unauthGuardService` instead. Again, `unauthGuardService` is subclassed to provide the client-specific redirection route. Note that `launchGuardService` calls `unauthGuardService` to make sure that the private mode check is done first. Otherwise, route guards are called asynchronously, which could cause problems if `unauthGuardService` is called in private mode.

* related jslib changes: https://github.com/bitwarden/jslib/pull/401

### CODE REVIEW

**Comment by eliykat** [comment]
jslib bumped & ready for merge

File: src/popup/services/unauth-guard.service.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+} from '@angular/router';
+
+import { UserService } from 'jslib-common/abstractions/user.service';
+import { VaultTimeoutService } from 'jslib-common/abstractions/vaultTimeout.service';
+
+import { UnauthGuardService as BaseUnauthGuardService } from 'jslib-angular/services/unauth-guard.service';
+
+@Injectable()
+export class UnauthGuardService extends BaseUnauthGuardService {
+    constructor(vaultTimeoutService: VaultTimeoutService, userService: UserService,
```
Comment (Hinton) [inline src/popup/services/unauth-guard.service.ts:13]: We could omit the constructor here?

File: src/popup/services/lock-guard.service.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+
+import { UserService } from 'jslib-common/abstractions/user.service';
+import { VaultTimeoutService } from 'jslib-common/abstractions/vaultTimeout.service';
+import { LockGuardService as BaseLockGuardService } from 'jslib-angular/services/lock-guard.service';
+
+@Injectable()
+export class LockGuardService extends BaseLockGuardService {
+    constructor(vaultTimeoutService: VaultTimeoutService, userService: UserService,
```
Comment (Hinton) [inline src/popup/services/lock-guard.service.ts:10]: We could omit the constructor here?

File: src/popup/services/unauth-guard.service.ts (hunk 1 above)
Comment (eliykat) [reply src/popup/services/unauth-guard.service.ts]: Good point, fixed!

File: src/popup/services/lock-guard.service.ts (hunk 2 above)
Comment (eliykat) [reply src/popup/services/lock-guard.service.ts]: Fixed

**Review by Hinton — APPROVED** [review]

**Pull request merged**